### ISSUE TITLE
entry to string function

### ISSUE THREAD

**Issue by goatshriek** [goatshriek/stumpless#448]
It is often useful to be able to obtain a string representation of a given value either for user interface hints, troubleshooting, or operations like value comparison. Such a function is currently implemented for some stumpless data types (like the version, param, and element structs) but not for others, including the entry struct.

Note that this function is _not_ intended for use in the formatting of a log message as this output may be subtly different, for example when escape codes are needed. Rather, it is meant to give an as-is representation of the entry itself (all fields and elements).

Check out the [Contributing Guidelines](https://github.com/goatshriek/stumpless/blob/latest/docs/CONTRIBUTING.md) and the [development guide](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md) for the basics on working with stumpless and submitting changes.

## General Approach
There are a few details left out of the following approach, for you to fill in as you encounter them. If you find you need help, please ask here or on the [project gitter](https://gitter.im/stumpless/community) and someone can help you get past the stumbling block. You are also free to take a different approach if you feel it is better - this is merely meant as a starting point.

First, read [this specific section](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions) of the development guide devoted to adding new functions. It provides an outline of the work you will need to do beyond the implementation and testing of the function.

Next, add a declaration for your function in the `include/stumpless/entry.h` header file. Name the function `stumpless_entry_to_string` and document its behavior in the same doxygen format as the others in the same header. Make sure to cover the following points:
 * the general format of the output. This should be each field name (procid, app_name, msgid, etc.), an equals sign, and the value in double quotes, all separated by commas. For the elements, use the `stumpless_element_to_string` function to get the output for each element. For example: `time="2003-10-11T22:14:15.003Z", prival="165", hostname="example.com", app_name="test-suite", procid="6", msgid="yellow", element-1=[param="val"], message="test message"`
 * the supplied argument should be a `const` pointer
 * explicitly state that the result must be freed when it is no longer needed

Now you are ready to add the matching implementation for the function in `src/entry.c`. Use the code for existing functions in this file as examples. You will also need to use `stumpless_element_to_string` to get the string representation of any elements in the entry. Be sure that there is no memory leak in your result, and try to avoid expensive operations like `sprintf` - you should be able to do this with `memcpy` calls and assignments, and perhaps `strlen` calls. You may consider refactoring some common portions of `stumpless_element_to_string` into separate private functions that can be used in both places for efficiency, but don't worry if you don't want to go this far - it can be turned into a follow-on first issue for others to tackle.

Finally, be sure to add tests for the new function. These should go in `test/function/entry.cpp` and should achieve full coverage, including for memory allocation failure and NULL argument cases. Again, there are already similar test cases for other functions in this file that you can look at to get a general idea of how the tests should be structured.

**Comment by Master9090rgb** [comment]
Hello Sir, I am new to the field of Open Source Contribution,I am willing to work on this issue if you can guide me a little on how to contribute to this. 
My Tech Stack is C++,Python,Django 
Thank you

**Comment by goatshriek** [comment]
Welcome! Have a look through the contributing guidelines and development guides linked in the issue, which are intended to be an accessible entry point for newcomers. Specifically, [this section](https://github.com/goatshriek/stumpless/blob/latest/docs/CONTRIBUTING.md#i-dont-know-where-to-start) lays out the most basic steps.

If you aren't super familiar with C and this is your first contribution, I do recommend one of the other good first issues that are smaller changes. This one will deal heavily with dynamic memory and strings, which can be tough to deal with coming from Python or even C++ in many cases. #445 may be a good fit, as it will only require a bugfix to existing C code, and the test suites are written in C++ which will be more familiar to you.

**Comment by Master9090rgb** [comment]
Thank You for the reply i will look through all the resources provided and will try to do my best

**Comment by KlowdfurrRad** [comment]
I would like to work on this issue.

**Comment by junyeong0619** [comment]
i made function to this for my first issue. before i submit this code, I really want to say sorry to stumpless's owner that's why i started write code about one and half years ago. Can i submit this?

**Comment by goatshriek** [comment]
Not a problem at all, thanks for contributing!

**Comment by junyeong0619** [comment]
> Not a problem at all, thanks for contributing!

i got a problem with registering stumpless_entry_to_string() function. Where can i add this function? In entry.h I can see only STUMPLESS_PUBLIC_FUNCTION define

**Comment by goatshriek** [comment]
Yes, this is the correct thing to add to the function declaration to expose it. This is explained in more detail in the documentation for [adding new functions](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions).

**Comment by junyeong0619** [comment]
> Yes, this is the correct thing to add to the function declaration to expose it. This is explained in more detail in the documentation for [adding new functions](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions).

I tried to make it in position private/entry.h and inserted malloc failure test is this right?

**Comment by junyeong0619** [comment]
> Yes, this is the correct thing to add to the function declaration to expose it. This is explained in more detail in the documentation for [adding new functions](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions).

And after i move my function definition to private/entry.h. The entry.cpp test cannot find my function. I think it is because of the entry.cpp include stumpless/entry.h

**Comment by junyeong0619** [comment]
> Yes, this is the correct thing to add to the function declaration to expose it. This is explained in more detail in the documentation for [adding new functions](https://github.com/goatshriek/stumpless/blob/latest/docs/development.md#adding-new-functions).

Can I ask why all of the tests failed in codecov and sonar? how can i test it?

**Comment by goatshriek** [comment]
Resolved by #501.
